.spin/cmds.py

```diff
@@ -81,3 +81,68 @@ def sdist(pyproject_build_args):
     sdist = os.path.join('dist', built_line.replace('Successfully built ', ''))
     print(f"Validating {sdist}...")
     spin.util.run(["tools/check_sdist.py", sdist])
+
+
+@click.option(
+    "--test-modified",
+    is_flag=True,
+    default=None,
+    help="Test only modified submodules",
+)
+@click.option(
+    "--test-modified-importers",
+    is_flag=True,
+    default=None,
+    help="Test all submodules that import changed submodules.",
+)
+@spin.cmds.meson.build_dir_option
+@spin.util.extend_command(spin.cmds.meson.test)
+def test(
+    *,
+    parent_callback,
+    build_dir,
+    test_modified=False,
+    test_modified_importers=False,
+    **kwargs,
+):
+    if test_modified or test_modified_importers:
+        sys.path.insert(0, 'tools/')
+        import module_dependencies
+
+        # Ensure spin-built version of skimage is accessible
+        p = spin.cmds.meson._set_pythonpath(build_dir, quiet=True)
+        sys.path.insert(0, p)
+
+        pkg_mods = module_dependencies._pkg_modules()
+
+        p = spin.util.run(
+            ['git', 'diff', 'main', '--stat', '--name-only'], output=False, echo=False
+        )
+        if p.returncode != 0:
+            raise (click.ClickException('Could not git-diff against main'))
+
+        git_diff = p.stdout.decode('utf-8')
+        changed_modules = {mod for mod in pkg_mods if mod.replace('.', '/') in git_diff}
+
+        if test_modified:
+            to_test = changed_modules
+        else:
+            to_test = set()
+
+        if test_modified_importers:
+            importers_of_modified = (
+                set(module_dependencies.modules_dependent_on(changed_modules))
+                - changed_modules
+            )
+            to_test = to_test | importers_of_modified
+
+        pytest_args = kwargs.get('pytest_args', ())
+        if "--pyargs" in pytest_args:
```

Comment:
The changes here now have somewhat major conflict with the recently introduced implementation for the `--doctest` option (#7786). Both override `--pyargs`.